db/cache_mutation_reader.hh

```diff
@@ -198,13 +198,13 @@ class cache_mutation_reader final : public mutation_reader::impl {
     }
 
     gc_clock::time_point get_read_time() {
-        return _read_context.tombstone_gc_state() ? gc_clock::now() : gc_clock::time_point::min();
+        return _read_context.gc_before_getter() ? gc_clock::now() : gc_clock::time_point::min();
     }
 
     gc_clock::time_point get_gc_before(const schema& schema, dht::decorated_key dk, const gc_clock::time_point query_time) {
-        auto gc_state = _read_context.tombstone_gc_state();
-        if (gc_state) {
-            return gc_state->with_commitlog_check_disabled().get_gc_before_for_key(schema.shared_from_this(), dk, query_time);
+        auto gc_before_getter = _read_context.gc_before_getter();
+        if (gc_before_getter) {
+            return gc_before_getter->with_commitlog_check_disabled().get_gc_before_for_key(schema.shared_from_this(), dk, query_time);
         }
```

Comment:
I hope there are no deep copies here. tombstone_gc_state has an std::function which might or might not have an allocated component, and here we are copying it around like there's no tomorrow.

Function objects are great when you can store them somewhere and repeatedly invoke, amortizing the cost, but here we're just creating a new one, calling it once, and throwing it away.

Please check if perf-simple-query regresses.

Reply:
After this PR, there is no more deep copy, by the end of the PR, `gc_before_getter` has just two pointers and a `size_t` and that is all.